File: leetcode/201-300/292-nim-game/nim_game.py
"""Oracle mirror for LeetCode 292 — same algorithm as nim_game.kara.

Losing positions are exactly the multiples of 4: from a multiple every move of
1, 2 or 3 lands on a non-multiple, and from a non-multiple you can always take
n % 4 and hand one back.
"""


def can_win_nim(n):
    return (n % 4) != 0


def report(n):
    print(f"n={n} -> {str(can_win_nim(n)).lower()}")


def main():
    for n in range(1, 13):
        report(n)
    report(2147483647)
    report(2147483644)
    report(0)


if __name__ == "__main__":
    main()

File: leetcode/201-300/292-nim-game/bench/nimgame.rs
// Benchmark twin for LeetCode #292 — same algorithm as nimgame.kara.

const N: usize = 20_000_000;

fn main() {
    let mut win = vec![false; N + 1];
    win[0] = false;
    for i in 1..=N {
        let mut w = false;
        for take in 1..=3usize {
            if i >= take && !win[i - take] {
                w = true;
            }
        }
        win[i] = w;
    }
    let mut losing: i64 = 0;
    let mut checksum: i64 = 0;
    for i in 0..=N {
        if !win[i] {
            losing += 1;
            checksum = (checksum * 31 + i as i64) % 1_000_000_007;
        }
    }
    println!("losing {} checksum {}", losing, checksum);
}
